src/lib/fetchGuildApi.ts

```diff
@@ -0,0 +1,171 @@
+ * @throws If `pathname` starts or ends with a slash.
+ * @throws If the API does not respond with JSON.
+ * @throws If the response JSON cannot be parsed.
+ */
+export const fetchGuildApi = async <Data = object, Error = ErrorLike>(
+  pathname: string,
+  requestInit?: RequestInit,
+): Promise<FetchResult<Data, Error>> => {
+  if (pathname.startsWith("/")) {
+    throw new Error("`pathname` must not start with slash");
+  }
+  if (pathname.endsWith("/")) {
+    throw new Error("`pathname` must not end with slash");
+  }
+
+  const url = new URL(`api/${pathname}`, env.NEXT_PUBLIC_API);
+  const response = await fetch(url, {
+    ...requestInit,
+    headers: {
+      ...(requestInit?.body && { "Content-Type": "application/json" }),
+      ...requestInit?.headers,
+    },
+  });
+
+  const contentType = response.headers.get("content-type");
+
+  if (!contentType?.includes("application/json")) {
+    throw new Error("Guild API failed respond with json");
+  }
+
+  logger.info("\n", url.toString(), response.status);
+
+  let json: unknown;
+  try {
+    json = await response.json();
+  } catch {
+    throw new Error("Failed to parse json from response");
+  }
+
+  logger.info(json, "\n");
+
+  if (!response.ok) {
+    return {
+      status: "error",
+      data: json as Error,
+      response,
+    };
+  }
+  return {
+    status: "success",
+    data: json as Data,
+    response,
+  };
+};
+
+/**
+ * Fetcher used for creating authenticated requests to the v3 backend API.
+ *
+ * This function extends `fetchGuildApi` by automatically including an authentication
+ * token in the request headers. The token is retrieved either server-side or client-side
+ * depending on the runtime environment.
+ *
+ * @template Data - The expected shape of the success response data.
+ * @template Error - The expected shape of the error response data.
+ *
+ * @param pathname - The API endpoint to fetch, without leading or trailing slashes.
+ * @param requestInit - Optional configuration for the fetch request.
+ *
+ * @returns A promise resolving to the fetch result,
+ * containing `status`, `data`, and `response`.
+ *
+ * @example
+ * ```ts
+ * const { data, status, response } = await fetchGuildApiAuth<{ user: string }>('user', {
+ *  method: "POST",
+ *  body: JSON.stringify({name: string}),
+ * });
+ *
+ * if (status === 'error') {
+ *   assert(!response.ok);
+ *   console.error(data.message);
+ * } else {
+ *   console.log(data.user);
+ * }
+ * ```
+ *
+ * @throws If `pathname` starts or ends with a slash.
+ * @throws If the API does not respond with JSON.
+ * @throws If the response JSON cannot be parsed.
+ * @throws If the JWT token cannot be retrieved.
+ */
+export const fetchGuildApiAuth = async <Data = object, Error = ErrorLike>(
+  ...[pathname, requestInit = {}]: Parameters<typeof fetchGuildApi>
+): Promise<FetchResult<Data, Error>> => {
+  const token = isServer
+    ? await tryGetToken()
+    : getCookieClientSide(GUILD_AUTH_COOKIE_NAME);
+  if (!token) {
```

Comment:
What happens when we need to send fetch requests to an endpoint both with and without auth? E.g. when we fetch the requirements/rewards for a guild? In that case, it isn't relevant if we send the token or not, we just want to display the roles which the user can see. We shouldn't throw in that case. 🤔 Or do we want to handle these requests separately? We should brainstorm about this!

Reply:
The `Auth` fetcher has been merged with the main fetcher. Currently the token error is not propagated. This could lead to the limitation of not being able to request something as an unauthenticated user.